test/replication-luatest/bootstrap_strategy_test.lua

```diff
@@ -455,3 +457,240 @@ g_config_fail.test_bad_uri_or_uuid = function()
+        old_tuple = new_tuple
+        new_tuple = old_tuple:update{{'+', 4, 1}}
+        box.space._schema:replace(new_tuple)
+        t.assert_equals(box.space._schema:get{'bootstrap_leader_uuid'},
+                        new_tuple, 'Last write wins by replica id - new tuple')
+
+        local ballot_uuid
+        local watcher = box.watch('internal.ballot', function(_, ballot)
+                local ballot_key = box.iproto.key.BALLOT
+                local uuid_key = box.iproto.ballot_key.BOOTSTRAP_LEADER_UUID
+                ballot_uuid = ballot[ballot_key][uuid_key]
+        end)
+        t.helpers.retrying({}, function()
+            t.assert_equals(ballot_uuid, new_tuple[2],
+                            'Ballot stores correct uuid')
+        end)
+        old_tuple = new_tuple
+        new_tuple = old_tuple:update{{'=', 2, uuid2}, {'-', 3, 1}}
+        box.space._schema:replace(new_tuple)
+        t.assert_equals(ballot_uuid, old_tuple[2],
+                        "Ballot isn't updated if the tuple is rejected")
+        new_tuple = new_tuple:update{{'+', 3, 1}}
+        box.space._schema:replace(new_tuple)
+        t.helpers.retrying({}, function()
+            t.assert_equals(ballot_uuid, new_tuple[2],
+                            'Ballot updates the uuid')
+        end)
+        old_tuple = new_tuple
+        new_tuple = new_tuple:update{{'=', 2, uuid3}}
+        box.begin()
+        box.space._schema:replace(new_tuple)
+        local new_uuid = ballot_uuid
+        box.commit()
+        t.assert_equals(new_uuid, old_tuple[2],
+                        "Ballot isn't updated before commit")
+        t.helpers.retrying({}, function()
+            t.assert_equals(ballot_uuid, new_tuple[2],
+                            "Ballot is updated on commit")
+        end)
+        watcher:unregister()
+    end)
+end
+
+local function assert_not_booted(server)
+    local logfile = fio.pathjoin(server.workdir, server.alias .. '.log')
+    t.helpers.retrying({}, function()
+        t.assert_equals(server:grep_log('ready to accept requests', nil,
+                                        {filename = logfile}), nil,
+                        server.alias .. 'does not bootstrap')
+    end)
+end
+
+g_supervised.before_test('test_wait_for_bootstrap_leader', function(cg)
+    cg.replica_set = replica_set:new{}
+    local box_cfg = {
+        bootstrap_strategy = 'supervised',
+        replication_timeout = 0.1,
+        replication = {
+            server.build_listen_uri('server1'),
+            server.build_listen_uri('server2'),
+        },
+        replication_connect_timeout = 1000,
+    }
+    for i = 1, 2 do
+        local alias = 'server' .. i
+        cg[alias] = cg.replica_set:build_and_add_server {
+            alias = alias,
+            box_cfg = box_cfg,
+        }
+    end
+end)
+
+g_supervised.test_wait_for_bootstrap_leader = function(cg)
+    cg.replica_set:start{wait_until_ready = false}
+    fiber.sleep(0.4)
```

Comment:
You might need to add a comment why 0.4.

Reply:
Thanks for pointing this out!
I decided to rewrite this test to rely less on timeout:
```diff
diff --git a/test/replication-luatest/bootstrap_strategy_test.lua b/test/replication-luatest/bootstrap_strategy_test.lua
index b0e94ec10..53d455c38 100644
--- a/test/replication-luatest/bootstrap_strategy_test.lua
+++ b/test/replication-luatest/bootstrap_strategy_test.lua
@@ -630,7 +630,7 @@ end
 
 g_supervised.before_test('test_wait_for_bootstrap_leader', function(cg)
     cg.replica_set = replica_set:new{}
-    local box_cfg = {
+    cg.box_cfg = {
         bootstrap_strategy = 'supervised',
         replication_timeout = 0.1,
         replication = {
@@ -643,20 +643,37 @@ g_supervised.before_test('test_wait_for_bootstrap_leader', function(cg)
         local alias = 'server' .. i
         cg[alias] = cg.replica_set:build_and_add_server{
             alias = alias,
-            box_cfg = box_cfg,
+            box_cfg = cg.box_cfg,
         }
     end
 end)
 
+local function wait_master_is_seen(replica, master, rs_id)
+    local addr = server.build_listen_uri(master.alias, rs_id)
+    local logfile = fio.pathjoin(replica.workdir, replica.alias .. '.log')
+    local query = string.format('remote master .* at unix/:%s',
+                                addr:gsub('%-', '%%-'))
+    t.helpers.retrying({}, function()
+        t.assert(replica:grep_log(query, nil, {filename = logfile}),
+                 replica.alias .. ' sees ' .. addr)
+    end)
+end
+
 g_supervised.test_wait_for_bootstrap_leader = function(cg)
     cg.replica_set:start{wait_until_ready = false}
-    fiber.sleep(0.4)
+
+    wait_master_is_seen(cg.server1, cg.server1, cg.replica_set.id)
+    wait_master_is_seen(cg.server1, cg.server2, cg.replica_set.id)
+    wait_master_is_seen(cg.server2, cg.server1, cg.replica_set.id)
+    wait_master_is_seen(cg.server2, cg.server2, cg.replica_set.id)
+
+    fiber.sleep(cg.box_cfg.replication_timeout)
+
     assert_not_booted(cg.server1)
     assert_not_booted(cg.server2)
 end

```